pkg/kubelet/cm/memorymanager/memory_manager.go

```diff
@@ -193,6 +193,17 @@ func (m *manager) AddContainer(pod *v1.Pod, container *v1.Container, containerID
 	defer m.Unlock()
 
 	m.containerMap.Add(string(pod.UID), container.Name, containerID)
+
+	// we know that init containers start in the same order as they present
+	// under the pod.spec.initContainers, so we delete all init containers until we
+	// reach one with the same name as the current one
+	for _, initContainer := range pod.Spec.InitContainers {
+		if initContainer.Name == container.Name {
+			break
+		}
+
+		m.policyRemoveContainerByRef(string(pod.UID), initContainer.Name)
+	}
```

Comment:
Why is this needed exactly? Why can't we just wait for `RemoveContainer` to clean this up for all containers? As it stands, things will be removed for all init containers (except that last one), which feels a little awkward (especially since it's not clear why it's necessary to remove this state here).

Reply:
The `RemoveContainer` is a redundant method that we do not really use under the container manager, see - https://github.com/kubernetes/kubernetes/pull/90377. So in general we clean things only during calls to
`GetTopologyHints`, `Allocate` or `AddContainer`

Reply:
Would it be possible to make an explicit call to `removeStaleState()` here instead?  I know it will end up doing a bit more redundant work, but at least it will be intuitive as to why we might be calling something like that here. As it stands it still feels awkward to just explicitly be removing some containers here in the way you are doing.

Reply:
The problem that `removeStaleState()` remove only containers for dead pods.
```
// Build a list of (podUID, containerName) pairs for all containers in all active Pods.
	activeContainers := make(map[string]map[string]struct{})
	for _, pod := range activePods {
		activeContainers[string(pod.UID)] = make(map[string]struct{})
		for _, container := range append(pod.Spec.InitContainers, pod.Spec.Containers...) {
			activeContainers[string(pod.UID)][container.Name] = struct{}{}
		}
	}
```
it does not really filter terminated containers. Do you want to refactor it?

Reply:
Dsicussed offline, and decided this is OK, so long as we add a more descriptive comment. Something like:
```
	// Since we know that each init container always runs to completion before
	// the next container starts, we can safely remove references to any previously
	// started init containers. This will free up the memory from these init containers
	// for use in other pods. If the current container happens to be an init container,
	// we skip deletion of it until the next container is added, and this is called again.
```